### PR TITLE
Introduction of metric filtering

### PR DESCRIPTION
**Pull request by utlemming**
Currently the agent collects a whole bunch of useful information that DigitalOcean is not consuming yet. This change filters out the metrics that are used. It is hoped that in the future the reporting server will ask for the metric it wants.
 
Changes:
- dropped reporting on pseudo devices and filesystems
- added regexp filters

### CODE REVIEW

File: collector/filter.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/digitalocean/do-agent/metrics"
+)
+
+// Filters is used to limit collection of metrics.
+type Filters struct {
+	IncludeAll bool
+	Regexps    []*regexp.Regexp
+}
+
+// ifIncluded call r.Update if the metric should be included.
```
Comment (cagedmantis) [inline collector/filter.go]: s/ifIncluded/UpdateIfIncluded/g

File: collector/filter.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+func (f *Filters) UpdateIfIncluded(r metrics.Reporter, ref metrics.MetricRef, value float64, labelValues ...string) {
```
Comment (cagedmantis) [inline collector/filter.go]: Would be great to have unit tests for this function.

File: collector/file_system.go [hunk 3]
```diff
@@ -27,10 +28,45 @@ const (
 	fsSystem = "filesystem"
 )
 
+// excludedPsudeoFSs are psudeo filesystems that are excluded from metrics.
+var excludedDevices = []string{
```
Comment (cagedmantis) [inline collector/file_system.go]: This is awesome.

File: cmd/do-agent/main.go [hunk 4]
```diff
@@ -50,6 +51,37 @@ var (
 	debugUpdateURL     = envflag.String("DO_AGENT_UPDATE_URL", update.RepoURL, "Override Update URL")
 	debugLocalRepoPath = envflag.String("DO_AGENT_REPO_PATH", update.RepoLocalStore, "Override Local repository path")
 	pluginPath         = envflag.String("DO_AGENT_PLUGIN_PATH", defaultPluginPath, "Override plugin path")
+
+	// By default, only collect these metrics _and_ any plugins metrics. In a future version of
+	// the agent, the server will be requesting the metrics to gather.
+	defaultMetrics = map[string]collector.Filters{
+		"cpu": collector.Filters{Regexps: []*regexp.Regexp{
+			regexp.MustCompile("cpu_cpu.*"),
```
Comment (cagedmantis) [inline cmd/do-agent/main.go]: Would love to see tests for the regex's used.

File: collector/file_system.go [hunk 5]
```diff
@@ -27,10 +28,45 @@ const (
 	fsSystem = "filesystem"
 )
 
+// excludedPsudeoFSs are psudeo filesystems that are excluded from metrics.
```
Comment (cagedmantis) [inline collector/file_system.go]: s/excludedPsudeoFSs/excludedDevices/g

**Comment by utlemming** [comment]
I'm going to merge this and then pull in some unit tests for the filter with the next cut.

**Pull request merged**